server/src/test/java/org/elasticsearch/search/aggregations/bucket/MergingBucketsDeferringCollectorTests.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.bucket;
+
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.NumericDocValuesField;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.IndexWriterConfig;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.search.aggregations.AggregatorTestCase;
+import org.elasticsearch.search.aggregations.BucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.bucket.MergingBucketsDeferringCollector;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.lucene.util.LuceneTestCase.newDirectory;
+import static org.mockito.Mockito.when;
+
+public class MergingBucketsDeferringCollectorTests extends AggregatorTestCase {
+
+    /**
+     * Usually all documents get collected into ordinal 0 unless they are part of a sub aggregation
+     * @return a query that collects the i'th document into bucket ordinal i
+     */
+    private Query getQueryToCollectIntoDifferentOrdinals() {
```

Comment:
I think it'd be a little cleaner to do this by wrapping the collector that you pass to `indexSearcher.search` and just us `MatchAllDocsQuery` instead.

Reply:
Fixed! I overrode one of the methods in `MergingBucketsDeferringCollector` and now it is a lot cleaner. Is this what you had in mind? 

I'm not sure if wrapping the bucket collector directly would work, since the deferring collector stores the bucket ordinal before it calls the bucket collector's `collect` method.

Reply:
That isn't quite what I was thinking. I think part of the problem is that I'm sort of stuck in a "do it like aggs do" mindset. In that mindset there are two collectors and the `MergingBucketsDeferringCollector`. One collector emulates the outer aggregator and calls to the `MergingBucketsDeferringCollector`'s collect method, calling `merge` in some funny shape. And the other bucket emulates the inner aggregation and is just called by the aggregator.

I've got half of a patch on my laptop that does this but I'm probably going to stop for the day. I'll try and post it this weekend.

Reply:
I was thinking something like [this](https://gist.github.com/nik9000/8317f6ae3baa3667db125e7f93ca3abe). There is a collector that just counts and a collector the distributes bucket ords and merges. And the `MergingBucketsDeferringCollector` sits between them.

Two neat things:
1. I think I found a bug! I don't think it actually comes up in production because you have to throw away buckets using the merge method while collecting buckets and I think we only do that in the `rare_terms` aggregators and they only merge after all the collections are done.
2. My test only really covers the variable-width histogram style of merging, not the rare_terms style. But I think that is pretty ok. Its is the harder style.

Reply:
Sorry for my delayed response, I had a very hectic start of week!

Very cool, that is definitely a lot cleaner, thanks for the patch! I've added it and filed a corresponding bug.
That's a strange bug by the way...

Reply:
> Sorry for my delayed response, I had a very hectic start of week!

Its cool! Thanks for getting back to me!

> That's a strange bug by the way...

It's sneaky! I'm hoping we really don't actually hit it. But it is the kind of thing that happens without unit tests, I think.